File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/IdentifierContext.java
package org.cqframework.cql.cql2elm;

import java.util.Objects;
import java.util.StringJoiner;
import org.cqframework.cql.elm.tracking.Trackable;
import org.hl7.elm.r1.IdentifierRef;

/**
 * Simple POJO using for identifier hider that maintains the identifier and Trackable type of the construct being evaluated.
 */
public class IdentifierContext {
    private final IdentifierRef identifierRef;
    private final Class<? extends Trackable> elementSubclass;

    public IdentifierContext(IdentifierRef identifierRef, Class<? extends Trackable> elementSubclass) {
        this.identifierRef = identifierRef;
        this.elementSubclass = elementSubclass;
    }

    public IdentifierRef getIdentifierRef() {
        return identifierRef;
    }

    public String getIdentifier() {
        return identifierRef.getName();
    }

    public Class<? extends Trackable> getTrackableSubclass() {
        return elementSubclass;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (other == null || getClass() != other.getClass()) {
            return false;
        }
        IdentifierContext that = (IdentifierContext) other;
        return Objects.equals(getIdentifier(), that.getIdentifier())
                && Objects.equals(elementSubclass, that.elementSubclass);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getIdentifier(), elementSubclass);
    }

    @Override
    public String toString() {
        return new StringJoiner(", ", IdentifierContext.class.getSimpleName() + "[", "]")
                .add("identifier='" + getIdentifier() + "'")
                .add("elementSubclass=" + elementSubclass)
                .toString();
    }
}

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/IdentifierContext.kt
package org.cqframework.cql.cql2elm

import java.util.*
import org.cqframework.cql.elm.tracking.Trackable

/**
 * Simple POJO using for identifier hider that maintains the identifier and Trackable type of the
 * construct being evaluated.
 */
class IdentifierContext(val identifier: String, val trackableSubclass: Class<out Trackable>?) {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as IdentifierContext
        return identifier == that.identifier && trackableSubclass == that.trackableSubclass
    }

    override fun hashCode(): Int {
        return Objects.hash(identifier, trackableSubclass)
    }

    override fun toString(): String {
        return StringJoiner(", ", IdentifierContext::class.java.simpleName + "[", "]")
            .add("identifier='$identifier'")
            .add("elementSubclass=$trackableSubclass")
            .toString()
    }
}
